ext/ddtrace_profiling_native_extension/http_transport.c

```diff
@@ -0,0 +1,281 @@
+    byte_slice_from_literal("ruby"),
+    (ddprof_ffi_Slice_tag) {.ptr = converted_tags, .len = tags_count},
+    endpoint_from(exporter_configuration)
+  );
+
+  xfree(converted_tags);
+
+  return exporter_result;
+}
+
+static VALUE handle_exporter_failure(ddprof_ffi_NewProfileExporterV3Result exporter_result) {
+  if (exporter_result.tag == DDPROF_FFI_NEW_PROFILE_EXPORTER_V3_RESULT_OK) return Qnil;
+
+  VALUE failure_details = rb_str_new((char *) exporter_result.err.ptr, exporter_result.err.len);
+
+  ddprof_ffi_NewProfileExporterV3Result_dtor(exporter_result);
+
+  return rb_ary_new_from_args(2, error_symbol, failure_details);
+}
+
+static ddprof_ffi_EndpointV3 endpoint_from(VALUE exporter_configuration) {
+  Check_Type(exporter_configuration, T_ARRAY);
+
+  ID working_mode = SYM2ID(rb_ary_entry(exporter_configuration, 0)); // SYMID verifies its input so we can do this safely
+
+  if (working_mode != agentless_id && working_mode != agent_id) {
+    rb_raise(rb_eArgError, "Failed to initialize transport: Unexpected working mode, expected :agentless or :agent");
+  }
+
+  if (working_mode == agentless_id) {
+    VALUE site = rb_ary_entry(exporter_configuration, 1);
+    VALUE api_key = rb_ary_entry(exporter_configuration, 2);
+    Check_Type(site, T_STRING);
+    Check_Type(api_key, T_STRING);
+
+    return ddprof_ffi_EndpointV3_agentless(byte_slice_from_ruby_string(site), byte_slice_from_ruby_string(api_key));
+  } else { // agent_id
+    VALUE base_url = rb_ary_entry(exporter_configuration, 1);
+    Check_Type(base_url, T_STRING);
+
+    return ddprof_ffi_EndpointV3_agent(byte_slice_from_ruby_string(base_url));
+  }
+}
+
+static void convert_tags(ddprof_ffi_Tag *converted_tags, long tags_count, VALUE tags_as_array) {
+  Check_Type(tags_as_array, T_ARRAY);
+
+  for (long i = 0; i < tags_count; i++) {
+    VALUE name_value_pair = rb_ary_entry(tags_as_array, i);
+    Check_Type(name_value_pair, T_ARRAY);
+
+    // Note: We can index the array without checking its size first because rb_ary_entry returns Qnil if out of bounds
+    VALUE tag_name = rb_ary_entry(name_value_pair, 0);
+    VALUE tag_value = rb_ary_entry(name_value_pair, 1);
+    Check_Type(tag_name, T_STRING);
+    Check_Type(tag_value, T_STRING);
+
+    converted_tags[i] = (ddprof_ffi_Tag) {
+      .name = byte_slice_from_ruby_string(tag_name),
+      .value = byte_slice_from_ruby_string(tag_value)
+    };
+  }
+}
+
+static VALUE _native_do_export(
+  VALUE self,
+  VALUE exporter_configuration,
+  VALUE upload_timeout_milliseconds,
+  VALUE start_timespec_seconds,
+  VALUE start_timespec_nanoseconds,
+  VALUE finish_timespec_seconds,
+  VALUE finish_timespec_nanoseconds,
+  VALUE pprof_file_name,
+  VALUE pprof_data,
+  VALUE code_provenance_file_name,
+  VALUE code_provenance_data,
+  VALUE tags_as_array
+) {
+  ddprof_ffi_NewProfileExporterV3Result exporter_result = create_exporter(exporter_configuration, tags_as_array);
```

Comment:
Do we need to build a new exporter because we can't yet set tags _after_ the creation of the exporter?